src/agent/debuglet.ts

```diff
@@ -106,6 +112,46 @@ const formatBreakpoints = function(
           .join('\n');
 };
 
+/**
+ * CachedPromise class is designed to support debug agent in Google Cloud
+ * Function (GCF). As GCF is a serverless environment and we wanted to make
+ * sure debug agent could always capture the snapshots. We designed this class
+ * so that GCF could wait the ListActiveBreakpoitns event completed before
+ * throttling CPU to 0. CachedPromise has two member functions.
+ * get() is to create or return an existing promise. resolve() will resolve
+ * the promise at appropriate code location (in our case is after
+ * listActiveBreakpoints executed).
+ */
+export class CachedPromise {
+  private promise: Promise<void>;
+  private promiseResolve: (() => void)|null;
+  private promiseResolvedTimestamp = -Infinity;
+  private refresh: number;
+  constructor(refresh: number) {
+    this.refresh = refresh;
+  }
+  get(): Promise<void> {
+    const diff = Date.now() - this.promiseResolvedTimestamp;
+    if (diff > this.refresh) {
+      this.promise = new Promise<void>((resolve) => {
+        this.promiseResolve = () => {
+          resolve();
+        };
```

Comment:
Making this a one-liner would be simpler.

```typescript
this.promiseResolve = resolve;
```

Reply:
Done.